### PR TITLE
membership details view

### PR DESCRIPTION
**Pull request by jay-meister**
Some things i still need to do:
#299 contains remaining todos for this PR

@rjmk could you please advise on:
- [x] tests are breaking due to the initial_value in the deactivated and reactivated parts of the member.js reducer. Initial value is supposed to store the value for that field currently on the database.
  Fixed
- [x] deactivate_member and reactivate_member actions seem to be calling twice, not too sure why.
  Due to my_details and form being in the root reducer.

### CODE REVIEW

**Comment by rjmk** [comment]
The problem is the type of the initial_state in the member reducer is a bit rubbishly defined (my fault).

You could change line 28 to read

`const reducer: Reducer<{ activation_status: {} }, Action> =`

and it should work

**Comment by jay-meister** [comment]
@rjmk didn't seem to fix the problem.

got it working if i changed lines 28 - 31

```
const initialState = { activation_status: {initial_value: 'activated'} }
const reducer: Reducer<{ activation_status: { initial_value: string } }, Action> =
  (member = initialState, { type, payload }) => {
```

AND

line 34 

```
        return { ...prepare_for_form(payload), ...member }
```

But i do not want to spread the remaining state from member after a fetch as we want all the values to be reset to the DB's values i think.  When i logged prepare_for_form(payload), activation_status was there so not sure why that fails..

**Comment by jay-meister** [comment]
if i don't add line 34 i get the following error message:
`
 30: const reducer: Reducer<{ activation_status: { initial_value: string } }, Action> =
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ property`activation_status`. Property not found in
 34:         return prepare_for_form(payload)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^ object literal
`

**Comment by rjmk** [comment]
Yeah, it's because the type system knows so little about what's going to be in the payload. There's probably a sensible way to fix that, but for now let's

``` js
const initialState = { activation_status: {} }
type State = typeof initialState

const reducer: Reducer<State, Action> =
  (member = initialState, { type, payload }) => {
    switch (type) {
      case FETCHED_MEMBER:
        return (prepare_for_form(payload): State)
```

**Comment by jay-meister** [comment]
nice. working now

**Comment by jay-meister** [comment]
I wonder if you can advise?

If the user adds a note when 'ending' their membership, i should hide all existing notes from them or only the ones that admin has made?
There is also a notes in the membership tab as requested by richard, I imagine we want to display any notes the user has previously made here but still hide the admins ones.

notes is just a string, so i'm not sure if i should proceed as it feels like it could get messy?

**Comment by rjmk** [comment]
@jackcarlisle Yes, you're quite right. I think it's going to be necessary to add a 'user_notes' column that can be concatenated with the admin notes for the admin

**Comment by jay-meister** [comment]
that sounds like it would be a lot easier. Is that a task for me or you? (i have no idea but happy to look into)

**Comment by rjmk** [comment]
Try looking at `api/models/Members.js` and see if it's obvious how to do. If not, I'm happy to do it

**Comment by jay-meister** [comment]
added 

```
    notes_user: {
      type: 'TEXT'
    },
```

and left `notes` for now.

**Comment by rjmk** [comment]
Nice. Might want to consider having a default of an empty string, but should be good

**Comment by jay-meister** [comment]
@rjmk think i have got around the need to introduce the new column and i think it is ready for you to look over.

still to do 
- [ ] Styling
- [ ] Decide what to show if member deactivates (awaiting Richards reply)
- [ ] Tidy up all 'returns' which i will do once you OK it.

**Comment by rjmk** [comment]
Some conflicts to resolve

**Comment by jay-meister** [comment]
think this is ready to be looked over again.

File: src/shared/redux/modules/member.js [hunk 1]
```diff
@@ -158,13 +162,19 @@ const fetch_user_url = () => '/api/account'
 
 const post_member_url = compose(make_user_url(post_user_url), prop('id'))
 
+const preppend_notes_if_there = mapObjIndexed((val, key, mem) =>
```
Comment (rjmk) [inline src/shared/redux/modules/member.js:165]: `prepend` with one 'p'

File: src/user/components/my_details/membership_details.js [hunk 2]
```diff
@@ -1,12 +1,32 @@
...
 
+import EndMembershipForm from './end_membership_form.js'
 import EditDetails from './edit_details.js'
 
-
-export default (props) =>
-  <div>
-    <div className='end-membership'>
-      <button>End Membership</button>
+export default (props) => {
```
Comment (rjmk) [inline src/user/components/my_details/membership_details.js:7]: No need for the braces and `return`

**Comment by rjmk** [comment]
OK

**Pull request merged**